library/ssl_client.c

```diff
@@ -258,7 +258,8 @@ static int ssl_write_supported_groups_ext(mbedtls_ssl_context *ssl,
     for (; *group_list != 0; group_list++) {
         MBEDTLS_SSL_DEBUG_MSG(1, ("got supported group(%04x)", *group_list));
 
-#if defined(PSA_WANT_ALG_ECDH)
+#if defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECDSA_C) || \
```

Comment:
This will break driver-only builds. TLS code should never check `ECDH_C or `ECDSA_C`, except in parts of the code guarded by `#if !defined(MBEDTLS_USE_PSA_CRYPTO)`.

I suggest `PSA_WANT_ALG_ECDH || PSA_WANT_ALG_ECDSA || MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED` for now, while waiting for #7612.

Reply:
The original code was guarded by `MBEDTLS_ECP_LIGHT` which is defined when `MBEDTLS_ECP_C` is defined.
Function `ssl_write_supported_groups_ext` is guarded by `MBEDTLS_ECDH_C` and `MBEDTLS_ECDSA_C`, so using these guards inside function should not cause any harm:
```
#if defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECDSA_C) || \
    defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED) || \
    (defined(MBEDTLS_SSL_PROTO_TLS1_3) && defined(PSA_WANT_ALG_FFDH))
```

Proposed solution won't help with failures when `config-suite-b.h` configuration is [used ](https://github.com/Mbed-TLS/mbedtls/pull/7627#issuecomment-1594545574) as `PSA_WANT_ALG_ECDH, PSA_WANT_ALG_ECDSA, MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED` are not defined in this configuration.

Reply:
Ok, I didn't notive this was pre-existing, so I withdraw my objection as long as the CI is happy. We'll come back to this in #7612 anyway.

Reply:
Still this condition seems not valid.
For example in `component_test_tls13_only_ephemeral_ffdh` we have `MBEDTLS_ECDH_C`, but `MBEDTLS_ECDSA_C` enabled. As a result EC curves are available and in e.g. `TLS 1.3: minimal feature sets - gnutls` EC curve is selected(29) and it can not be handled by `mbedtls_ssl_tls13_generate_and_write_dh_key_exchange`.

Reply:
Ah, good point again. This is a toughie. Our lives would be easier if we had done #7612 and #7613 first, but here we are.

Let me try a new suggestion:
```
#if (defined(MBEDTLS_SSL_PROTO_TLS1_2) &&
    (defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECDSA_C) || \
    defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED))) || \
    defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_SOME_EPHEMERAL_ENABLED)
static int ssl_write_supported_groups_ext(...)
{
    // [common beginning]
#if (defined(MBEDTLS_SSL_PROTO_TLS1_2) &&
    (defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECDSA_C) || \
    defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED))) || \
    (defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_SOME_EPHEMERAL_ENABLED) && \
     defined(PSA_WANT_ALG_ECDH))
    // [ECC curves]
#endif
#if defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_SOME_EPHEMERAL_ENABLED) &&\
     defined(PSA_WANT_ALG_FFDH)
    // [FFDH groups]
#endif
    // [common ending]
}
```
That is: we keep the existing ECDH || ECDSA || ECJPAKE but group them with `&& TLS 1.2`, and then add `|| TLS 1.3` with the proper `PSA_WANT` macros.

That keeps the TLS 1.2 part close to the existing while keeping the TLS 1.3 part clean. Do you think that would work or did I miss something again?

Reply:
PS: if we go that way, we'll probably want to define new macro for the TLS 1.2 part to avoid repeating it:
```c
#if defined(MBEDTLS_SSL_PROTO_TLS1_2) && \
    (defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECDSA_C) || \
    defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED))
#defined MBEDTLS_SSL_TLS1_2_SOME_ECC
#endif
```

Reply:
Checked suggested solution with mentioned test and it has passed.

Reply:
Phew :)